rand_distr/src/multinomial.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright 2018 Developers of the Rand project.
+// Copyright 2013 The Rust Project Developers.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! The multinomial distribution.
+
+use crate::{Binomial, Distribution};
+use num_traits::AsPrimitive;
+use rand::Rng;
+
+/// Error type returned from `Multinomial::new`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// There is a negative weight or Nan
+    ProbabilityNegative,
+    /// Sum overflows to inf
+    SumOverflow,
+    /// Sum is zero
+    SumZero,
+}
+
+impl core::fmt::Display for Error {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.write_str(match self {
+            Error::ProbabilityNegative => "One of the weights is negative or Nan",
+            Error::SumOverflow => "Sum of weights overflows to inf",
+            Error::SumZero => "Sum of weights is zero",
+        })
+    }
+}
+
+/// The [Multinomial](https://en.wikipedia.org/wiki/Multinomial_distribution) distribution `Multinomial(n, w)`.
+#[derive(Debug)]
+pub struct Multinomial {}
+
+impl Multinomial {
```

Comment:
This struct is just a stub used to construct the const/dyn variants? Then I'd prefer you just move `Multinomial::new_const` → `MultinomialConst::new` etc.; it's more predictable and one less import for common usage.

Reply:
My thought was that each distribution would have one common struct to construct all its variants.
Taking https://github.com/rust-random/rand/issues/494 into account, we might end up with a lot of distribution structs. Most people would never care about the exact type returned.
Having all the structs in one common doc page would probably do the same.